### ISSUE TITLE
Accessibility: NVDA reads entire list of suggestions in intelliSense rather than currently focused one

### ISSUE THREAD

**Issue by zersiax** [microsoft/vscode#92719]
- VSCode Version: 1.44 Insider
- OS Version: Windows 10 
build 19577
Steps to Reproduce:

1. Open Code, create a file you have a language extension installed for and bring up the suggestions panel by typing.
2. NVDA will read the entire list of suggestions, followed by the currently selected one, followed by "[index] of ]max]".
What follows is a bit of log output from NVDA after , in a flask-based Python file, the phrase "fr"  is typed.

```
IO - speech.speak (13:42:49.990) - MainThread (8652):
Speaking [LangChangeCommand ('en'), 'filter float for format from frozenset from_conf_import_settings from_trans_import_gettext from_trans_import_ngettext from_trans_import_pgettext from_trans_import_ugettext_lazy for', 'list']
IO - speech.speak (13:42:50.007) - MainThread (8652):
Speaking [LangChangeCommand ('en'), 'filter', '13 of 24']
```

Does this issue occur when all extensions are disabled?: Yes

@isidorn  I strongly suspect this might have to do with the way issue 10777 was resolved, could you look into this?
@jcsteh Does this have something to do with the way NVDA handles role-list?

**Comment by jcsteh** [comment]
I don't use VS Code, so I can't comment on specifics here. I can't think of
a case where NVDA would itself descend into lists and read descendants.
However, it's possible that Chromium is calculating the name from
descendants if the list itself gets focused, though that would surprise me
since I believe that wouldn't be spec compliant.

**Comment by isidorn** [comment]
@zersiax I can reproduce this issue, thanks for reprorting it. When did this exactly break, might you know? Why do you thing issue #10777 is related, I think you have a wrong issue number?
@leonardder might it be possible that there was a change in NVDA that broke this?

Please note that we changed the VS Code role to be `application` and that the `aria-selected` is now set a bit differently. However reverting any of those changes does not help here.
I am open for ideas.

**Comment by zersiax** [comment]
@isidorn my bad, I meant issue 90876. The number I gave initially is the issue on NVDA's side :)
So to clarify, when we went from treeviewItems to ListItems because it was causing trouble, do you know if this was already happening as well?

**Comment by isidorn** [comment]
@zersiax I have just verified and you are correct. The actual problem is that the `role=list`, once I change that to be `role=tree` all is good. The actual items having the `role=listitem` does not seem to be a problem.
I can of course make the change and revert our correct behavior of setting `role=list`. Though it would be more interesting if this can be fixed on the NVDA side and we remain with the correct role.

Otherwise we would use the role=tree only on windows so NVDA plays nice.

@leonardder @jcsteh let me know what you think. Thanks a lot!

**Comment by leonardder** [comment]
I think the problem is that you set the role to list, whereas it should be listbox. See https://www.w3.org/WAI/PF/aria/roles#listbox

**Comment by isidorn** [comment]
@leonardder you are correct. I just verified this indeed fixes the issue.
I have pushed a fix. @zersiax let me know if vscode insiders is good for you tomorrow. And thanks for creating this issue!

**Comment by isidorn** [comment]
Actually after reading more about the roles here https://www.w3.org/WAI/PF/aria/roles#listbox
The role `listbox`is not so fitting for our lists. The role `list` is better.

Though I do understand that `listbox` might be better just for inteliSense, however we do not respect the spec that the children of that element should be `option` since for us the children are `listitem`.
@leonardder so my question is should the children of inteliSense be set to the role `option`

**Comment by leonardder** [comment]
> Actually after reading more about the roles here https://www.w3.org/WAI/PF/aria/roles#listbox
> The role `listbox`is not so fitting for our lists. The role `list` is better.

Why do you think so? The role listbox IMO is the best choice when you have to deal with selecting one or more items in a list, whereas the list role is more for static lists. May be someone with more ARIA knowhow such as @marcozehe can help us out.
[Here](https://www.w3.org/TR/core-aam-1.2/) are the accessibility mappings. You can see t hat list maps to a readonly list, whereas listbox does not. Read-only here means that you can't select entries in the list.

> @leonardder so my question is should the children of inteliSense be set to the role `option`

IN the case of listbox, yes, sorry for missing that.

**Comment by isidorn** [comment]
Got it. So doing the following change: static tree will have role `tree` and children `treeitem`
static lists will have role `list` and children `listitem`
InteliSense will have role `listbox` and children `option`

**Comment by MarcoZehe** [comment]
@leonardder Yes, your assessment is correct. For an interactive list, where a user makes a choice, role "listbox" for the container and role "option" for the actual items are the correct roles to use. The actual thing getting focus, or being targeted by aria-activedescendant if that is used, is the visually highlighted element with role "option". That obviously needs to change with each press of the arrow keys, as the visual highlight also moves.

I believe the design pattern most closely resembling what you need is [combobox](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox), with either the [manual selection](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html) or [Selection with inline autocomplete](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-both.html) variants. The difference is that your editable aria is the code editor, not a single line search or such. But Twitter, Facebook, and others have examples where this is possible inline, so there is precedence for this kind of autocomplete.

**Comment by isidorn** [comment]
@MarcoZehe thanks a lot. We have just made the change to use `listbox` for our "interactive lists" or the inteliSense. You can try it out with vscode insdiers from Monday.

**Comment by isidorn** [comment]
@zersiax did you get a chance to try this out in [VS Code insiders](https://code.visualstudio.com/insiders/)? Asking because we would like to add a verified label to this issue if it got fixed for you? Thanks

**Comment by zersiax** [comment]
Hi :) Yep, I can see this is working now. The only thing that still makes intellisense a little cumbersome now is the issue that when the list gets dismissed or updated, the filename etc. gets read out by NVDA but I'm pretty sure there is already a separate issue for that somewhere :) This one's fixed, though, thanks @isidorn

**Comment by isidorn** [comment]
@zersiax Great, thanks adding verified label.
For the other issue we have this one tracking it https://github.com/microsoft/vscode/issues/89098 which is tracked on the NVDA side here https://github.com/nvaccess/nvda/issues/10887
